src/blob.rs

```diff
@@ -978,18 +981,24 @@ mod tests {
             .write_to(&mut buf, image::ImageFormat::Jpeg)
             .unwrap();
         let bytes = buf.into_inner();
-        let img_rotated = send_image_check_mediaquality(
-            Some("0"),
-            &bytes,
-            BALANCED_IMAGE_SIZE * 1800 / 2000,
-            BALANCED_IMAGE_SIZE,
-            0,
-            BALANCED_IMAGE_SIZE * 1800 / 2000,
-            BALANCED_IMAGE_SIZE,
-        )
-        .await
-        .unwrap();
-        assert_correct_rotation(&img_rotated);
+
+        // Do this in parallel to speed up the test a bit
```

Comment:
IMO parallelizing tests from the inside adds unnecessary complexity. Test runner already runs the tests in parallel.

Also I'm not even sure parallelizing with async tasks works when you run blocking tasks (computations) inside. For computations, like key generation, we spawn threads.

Reply:
On my machine with CPU locked to minimum speed (1,4GHz) (to get reproducible performance, otherwise it would start to throttle at some point), parallelizing this test "from the inside" reduces the time for `cargo test` from 51.8s to 46.3s and for `cargo nextest run` from 41.7s to 36.4s. Of course on CI there probably won't be as much performance improvement or maybe none at all, that's only for local builds. But there won't be a performance penalty on CI either since async tasks are cheap.

I tested like this:

<details>

Un-parallelize the test:
```diff
diff --git a/src/blob.rs b/src/blob.rs
index 276ac897..c0fa9d8e 100644
--- a/src/blob.rs
+++ b/src/blob.rs
@@ -982,13 +982,9 @@ async fn test_recode_image_2() {
             .unwrap();
         let bytes = buf.into_inner();
 
-        // Do this in parallel to speed up the test a bit
-        // (it still takes very long though)
-        let bytes2 = bytes.clone();
-        let join_handle = async_std::task::spawn(async move {
         let img_rotated = send_image_check_mediaquality(
             Some("0"),
-                &bytes2,
+            &bytes,
             BALANCED_IMAGE_SIZE * 1800 / 2000,
             BALANCED_IMAGE_SIZE,
             0,
@@ -998,7 +994,6 @@ async fn test_recode_image_2() {
         .await
         .unwrap();
         assert_correct_rotation(&img_rotated);
-        });
 
         let img_rotated = send_image_check_mediaquality(
             Some("1"),
@@ -1012,8 +1007,6 @@ async fn test_recode_image_2() {
         .await
         .unwrap();
         assert_correct_rotation(&img_rotated);
-
-        join_handle.await;
     }
 
     #[async_std::test]
```

then run:

<pre><font color="#FF5555"><b>➜  </b></font><font color="#55FFFF"><b>deltachat-core-rust</b></font> <font color="#5555FF"><b>git:(</b></font><font color="#FF5555"><b>hoc/speedup-rust-tests</b></font><font color="#5555FF"><b>) </b></font><font color="#FFC700"><b>✗</b></font> cargo test --no-run; time cargo test -q; time cargo nextest run 2&gt;/dev/null; git stash push; cargo test --no-run; time cargo test -q; time cargo nextest run 2&gt;/dev/null
<font color="#55FF55"><b>    Finished</b></font> test [unoptimized] target(s) in 0.38s

running 549 tests
<font color="#00AA00">....................................................................................................</font> 100/549
<font color="#00AA00">....................................................................................................</font> 200/549
<font color="#00AA00">....................................................................................................</font> 300/549
<font color="#00AA00">....................................................................................................</font> 400/549
<font color="#00AA00">....................................................................................................</font> 500/549
<font color="#00AA00">.................................................</font>
test result: <font color="#00AA00">ok</font>. 549 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 42.00s


running 1 test
<font color="#00AA00">.</font>
test result: <font color="#00AA00">ok</font>. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 5.34s

cargo test -q  193,55s user 32,52s system 436% cpu 51,831 total
cargo nextest run 2&gt; /dev/null  179,79s user 11,83s system 459% cpu 41,727 total
Saved working directory and index state WIP on speedup-rust-tests: c699ca49 Speedup rust tests
<font color="#55FF55"><b>   Compiling</b></font> deltachat v1.77.0 (/home/jonathan/deltachat-android/jni/deltachat-core-rust)
<font color="#55FF55"><b>    Finished</b></font> test [unoptimized] target(s) in 1m 01s

running 549 tests
<font color="#00AA00">....................................................................................................</font> 100/549
<font color="#00AA00">....................................................................................................</font> 200/549
<font color="#00AA00">....................................................................................................</font> 300/549
<font color="#00AA00">....................................................................................................</font> 400/549
<font color="#00AA00">....................................................................................................</font> 500/549
<font color="#00AA00">.................................................</font>
test result: <font color="#00AA00">ok</font>. 549 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 36.91s


running 1 test
<font color="#00AA00">.</font>
test result: <font color="#00AA00">ok</font>. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 5.09s

cargo test -q  192,82s user 28,71s system 477% cpu 46,369 total
cargo nextest run 2&gt; /dev/null  179,59s user 12,06s system 525% cpu 36,498 total
</pre>


</details>



> Also I'm not even sure parallelizing with async tasks works when you run blocking tasks (computations) inside.

My understanding is that async_std spawns as many executor threads as there are CPUs.